validator/client/wait_for_activation.go

```diff
@@ -53,65 +52,41 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
 func (v *validator) internalWaitForActivation(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
 	ctx, span := trace.StartSpan(ctx, "validator.WaitForActivation")
 	defer span.End()
-
 	validatingKeys, err := v.keyManager.FetchValidatingPublicKeys(ctx)
 	if err != nil {
-		return errors.Wrap(err, "could not fetch validating keys")
+		return errors.Wrap(err, msgCouldNotFetchKeys)
 	}
 	if len(validatingKeys) == 0 {
 		log.Warn(msgNoKeysFetched)
-
-		ticker := time.NewTicker(keyRefetchPeriod)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-ticker.C:
-				validatingKeys, err = v.keyManager.FetchValidatingPublicKeys(ctx)
-				if err != nil {
-					return errors.Wrap(err, msgCouldNotFetchKeys)
-				}
-				if len(validatingKeys) == 0 {
-					log.Warn(msgNoKeysFetched)
-					continue
-				}
-			case <-ctx.Done():
-				log.Debug("Context closed, exiting fetching validating keys")
-				return ctx.Err()
-			}
-			break
-		}
 	}
-
-	req := &ethpb.ValidatorActivationRequest{
-		PublicKeys: bytesutil.FromBytes48Array(validatingKeys),
-	}
-	stream, err := v.validatorClient.WaitForActivation(ctx, req)
-	if err != nil {
-		tracing.AnnotateError(span, err)
-		attempts := streamAttempts(ctx)
-		log.WithError(err).WithField("attempts", attempts).
-			Error("Stream broken while waiting for activation. Reconnecting...")
-		// Reconnection attempt backoff, up to 60s.
-		time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
-		return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
-	}
-
-	if err = v.handleAccountsChanged(ctx, accountsChangedChan, &stream, span); err != nil {
-		return err
-	}
-
-	v.ticker = slots.NewSlotTicker(time.Unix(int64(v.genesisTime), 0), params.BeaconConfig().SecondsPerSlot)
-	return nil
-}
-
-func (v *validator) handleAccountsChanged(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte, stream *ethpb.BeaconNodeValidator_WaitForActivationClient, span *trace.Span) error {
+	// loop while there are no validator keys ...
 	for {
 		select {
+		case <-ctx.Done():
+			log.Debug("Context closed, exiting fetching validating keys")
+			return ctx.Err()
 		case <-accountsChangedChan:
-			// Accounts (keys) changed, restart the process.
+			// if the accounts changed try it again
 			return v.internalWaitForActivation(ctx, accountsChangedChan)
 		default:
-			res, err := (*stream).Recv()
+			if len(validatingKeys) == 0 {
+				continue
+			}
```

Comment:
do we need maybe some small sleep here? we are here in the endless for loop and if channels won't be filled for some longer period I think we can experience some CPU spike here?